Assets/Scripts/Gamestate/AIManager.cs

```diff
@@ -186,61 +227,78 @@ private IEnumerator LookForTargets()
-        Transform closestPlayer = null;
-        float closestDistance = ignoreAwareRadius;
-        foreach (var player in TrackedPlayers)
-        {
-            var targetDirection = player.AiTarget.transform.position - transform.position;
-            var hitDistance = targetDirection.magnitude;
-            var playerDistance = (player.transform.position - transform.position).magnitude;
-            // Is the player nearby?
-            if (playerDistance > ignoreAwareRadius)
-                continue;
-            if (playerDistance > autoAwareRadius)
-            {
-                // Is the tracked player in front of me? (viewable)
-                if (Vector3.Dot(transform.forward, targetDirection) < 0)
-                    continue;
-                // Is there a line of sight to a tracked player?
-                if (Physics.Raycast(transform.position, targetDirection, hitDistance - 0.1f, ignoreMask))
-                    continue;
-                // Is there another tracked player who is closer?
-                if (hitDistance > closestDistance)
-                    continue;
-            }
-
-            closestPlayer = player.AiTarget.transform;
-            closestDistance = hitDistance;
-            DestinationTarget = closestPlayer;
-            ShootingTarget = player.AiAimSpot;
-        }
+        var closestPlayer = FindClosestPlayer(out var closestDistance);
 
         if (closestPlayer == null)
         {
             ShootingTarget = null;
             if (DestinationTarget == null || !DestinationTarget || (!DestinationTarget.gameObject.GetComponent<PlayerManager>() && !DestinationTarget.gameObject.activeInHierarchy))
             {
-                var target = MatchController.Singleton.GetRandomActiveChip();
-                DestinationTarget = target;
+                DestinationTarget = MatchController.Singleton.GetRandomActiveChip();
             }
         }
         else
         {
+            // TODO do not default to shooting target??? what's this for?
             aiMovement.Target = ShootingTarget;
             var isStrafeDistance = closestDistance < 10f;
             aiMovement.enabled = isStrafeDistance;
+
+            if (ShootingTarget != null)
+            {
+                var closestExplodingBarrel = ExplodingBarrel.GetViableExplodingBarrel(ShootingTarget.position);
+                var isBarrelSafeToShoot = closestExplodingBarrel
+                    && Vector3.Distance(closestExplodingBarrel.transform.position, transform.position) > closestExplodingBarrel.Radius + 1;
+                if (isBarrelSafeToShoot)
+                {
+                    // TODO respond to the barrel actually exploding?
+                    // TODO the shooting target going missing seems to be the cause for 
+                    ShootingTarget = closestExplodingBarrel.transform;
+                    // TODO unsub?
+                    closestExplodingBarrel.GetComponent<HealthController>().onDeath += HandleExplodedBarrel;
```

Comment:
Maybe unsubscrive at the HandleExplodedBarrel function itself? Might be nice in order to minimize memory leakage

Reply:
Thought I did that already but must have forgor 💀

Reply:
Nope, I already did every single thing that the TODO comments around here were talking about. Removed the superfluous comments now 🙃